src/Interpreters/findGoodReordering.cpp

```diff
@@ -0,0 +1,246 @@
+{
+    /// returns numbers from 0 up to N, excluding N.
+    std::vector<size_t> indexes(N);
+    for (size_t i = 0; i < N; ++i)
+        indexes[i] = i;
+    return indexes;
+}
+
+static size_t removeElems(size_t n, std::vector<size_t> & elems_to_remove, Vector2D<Float64> & selectivities, std::vector<Float64> & selectivity_to_joined, std::vector<Float64> & self_selectivities, std::vector<Float64> & sizes, std::vector<size_t> & indexes)
+{
+    /// We do "removal" of indexes from the data structures used by greedy() by
+    ///     1) copying the element at the back of the data structure, at index n-1, to the index to be removed.
+    ///     2) decrementing n.
+    for (size_t i = 0; i < elems_to_remove.size(); ++i)
+    {
+        size_t idx_to_remove = elems_to_remove[i];
+
+        if (i > 0)
+            idx_to_remove = std::find(indexes.begin(), indexes.end(), idx_to_remove) - indexes.begin();
+
+        //node idx_to_remove is now a part of the join tree, update selectivities for the remaining nodes.
+        for (size_t k = 0; k < n; k++)
+        {
+            selectivity_to_joined[k] *= selectivities(idx_to_remove, k);
+        }
+
+        //move last node (index: n-1) to the place of selected node (index: idx_to_remove).
+        if (idx_to_remove != n-1)
+        {
+            for (size_t k = 0; k < n; k++)
+            {
+                selectivities(k, idx_to_remove) = selectivities(n-1, k);
+                selectivities(idx_to_remove, k) = selectivities(n-1, k);
+            }
+            selectivity_to_joined[idx_to_remove] = selectivity_to_joined[n-1];
+            self_selectivities[idx_to_remove] = self_selectivities[n-1];
+            sizes[idx_to_remove] = sizes[n-1];
+            indexes[idx_to_remove] = indexes[n-1];
+        }
+        n -= 1;
+    }
+    return n;
+}
+
+static Float64 calculateCost(Float64 size_left, Float64 size_right, Float64 num_cols_right, Float64 sel_left, Float64 sel_right)
+{
+    /// The constants here come from a linear regression fitted to predict runtimes of joins of two tables.
+    /// The regression fits the vector coefs to sum(X*coefs)=y
+    /// y = join runtime in microseconds
+    /// X = [    size_left,  size_right, size_left*num_cols_left, size_right*num_cols_right, size_left*sel_left, size_right*sel_right]
+    /// coefs = [0.01709865, 0.        , 0.                     , 0.01674002,                0.06704978,          0.3567766 ]
+    /// The regression was constrained to have non-negative coefficients. All examples in the training data used hash joins.
+    /// It got an R2 = 0.965 on the training data (which had 320 samples).
+    /// Inputs to the regression are clearly correlated, and coefficients may not map to any real characteristics of how joins work.
+    return size_left * 0.01709865 +
+           num_cols_right * size_right * 0.01674002 +
+           size_left * sel_left * 0.06704978 +
+           size_right * sel_right * 0.3567766;
```

Comment:
Could you please elaborate on how did you train the weights? What is the training data? Is it the same dataset used for benchmarking?

Reply:
The training data is completely independent of the benchmark.

I don't have the code in front of me atm, but from memory: 
  - I built tables with with UInt64 columns, in various sizes varying linearly from 1000 to 100M. 
  - They were constructed so that the result of an inner join between the two tables would have the size of the smallest of the two tables.
  - The non-primary keys were in random order
  - Then, for each combination of table sizes (a,b) I benchmarked joining the tables, varying certain dimensions
    - benchmarking inner_join(a,b) and inner_join(b,a)
    - projection of 0, 1 or 2 columns of the right table
    - projection of 0, 1 or 2 columns of the left table
    - a where clause with approx ~50% selectivity on the right table
    - a where clause with approx ~50% selectivity on the left table
    - joining on primary key vs joining on non-primary key
  - These dimensions were *not* varied in the dataset:
    - The join algorithm
    - Size of the data type of the joined column 
  - Some care was taken to make each dimension "equally represented", though it's debatable how to do this.

Finally a linear regression was fit to predict the runtime of the joins, based on these characteristics. The runtimes were measured on a build of 23.3, so this may be improvable.

I think there's some danger that these numbers could become an unmaintainable magic no one dares to touch. I don't have it ready, but this can be packaged into a runnable python script that prints out coefficients. Should I include this in the PR? Or maybe add it to a separate repo?

Reply:
I added code to fit coefficients in `utils/tune-join-coefficients/run.sh`